infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -66,96 +70,77 @@
+    private final long maxConflictNewSequencer;
 
     @Getter
     private final String conflictKeysCounterName = "sequencer.conflict-keys.size";
     @Getter
     private final String windowSizeName = "sequencer.cache.window";
+
     /**
      * The cache limited by size.
      * For a synchronous cache we are using a same-thread executor (Runnable::run)
      * https://github.com/ben-manes/caffeine/issues/90
      *
      * @param cacheSize cache size
      */
-
     public SequencerServerCache(int cacheSize, long maxConflictNewSequencer) {
         this.cacheSize = cacheSize;
 
-        cacheEntries = new PriorityQueue(cacheSize, Comparator.comparingLong
-                (conflict -> ((ConflictTxStream) conflict).txVersion));
         maxConflictWildcard = maxConflictNewSequencer;
         this.maxConflictNewSequencer = maxConflictNewSequencer;
-        conflictKeys = MeterRegistryProvider
+        conflictKeyMap = MeterRegistryProvider
                 .getInstance()
                 .map(registry ->
                         registry.gauge(conflictKeysCounterName, Collections.emptyList(),
-                                new HashMap<ConflictTxStream, Long>(), HashMap::size))
+                                new HashMap<ConflictTxStream, Long>(), Map::size))
                 .orElse(new HashMap<>());
         MeterRegistryProvider.getInstance().map(registry ->
                 Gauge.builder(windowSizeName,
-                        conflictKeys, HashMap::size).register(registry));
-
+                        conflictKeyMap, Map::size).register(registry));
     }
 
     /**
      * Returns the value associated with the {@code key} in this cache,
-     * or {@code null} if there is no cached value for the {@code key}.
+     * or {@code Address.NON_ADDRESS} if there is no cached value for the {@code key}.
      *
      * @param conflictKey conflict stream
      * @return global address
      */
     public Long get(ConflictTxStream conflictKey) {
-        return conflictKeys.getOrDefault(conflictKey, Address.NON_ADDRESS);
+        return conflictKeyMap.getOrDefault(conflictKey, Address.NON_ADDRESS);
     }
 
     /**
-     * The first address in the priority queue.
+     * The first (smallest) address in the cache.
      */
     public long firstAddress() {
-        if (cacheEntries.isEmpty()) {
-            return Address.NOT_FOUND;
-        }
-        return cacheEntries.peek().txVersion;
+        return firstAddress;
     }
 
   /**
-   * Invalidate the records with the minAddress. It could be one or multiple records
+   * Invalidate the records with the firstAddress. It could be one or multiple records
    *
    * @return the number of entries has been invalidated and removed from the cache.
    */
   private int invalidateSmallestTxVersion() {
-    ConflictTxStream firstEntry = cacheEntries.peek();
-    if (cacheEntries.size() == 0) {
-      return 0;
-    }
+      if (versionMap.isEmpty()) {
```

Comment:
I renamed the relevant methods as suggested, but evicting as a function of the number of elements is awkward since we really want to evict as a function of timestamps, and we don't know ahead of time how many entries we will end up needing to evict. If we insert K elements and the cache can't fit them all, then we might end up evicting anywhere from 1 to K versions, depending on how many conflict keys are associated with each.